File: sdks/python/src/zeroclaw_plugin_sdk/tools.py
"""Tools module — wraps the zeroclaw_tool_call host function.

Serializes a typed request dict to JSON, calls the ``zeroclaw_tool_call``
host function via Extism shared memory, and deserializes the JSON response.
"""

from __future__ import annotations

import json
from typing import Any, Dict

import extism_pdk as pdk


def tool_call(tool_name: str, arguments: Dict[str, Any] | None = None) -> str:
    """Call a tool by name with the given arguments.

    Args:
        tool_name: Name of the tool to invoke.
        arguments: JSON-serializable dict of arguments for the tool.
            Defaults to an empty dict if not provided.

    Returns:
        The output string from the tool on success.

    Raises:
        RuntimeError: If the host reports an error or the call fails.
    """
    request = json.dumps({
        "tool_name": tool_name,
        "arguments": arguments if arguments is not None else {},
    })
    raw = pdk.host_fn("zeroclaw_tool_call", request)
    response = json.loads(raw)
    if response.get("error"):
        raise RuntimeError(response["error"])
    if not response.get("success"):
        raise RuntimeError("tool call returned success=false")
    return response.get("output", "")

File: crates/zeroclaw-plugin-sdk/src/tools.rs
//! Tools module — wraps tool_call host function.
//!
//! Serializes a typed request to JSON, calls the `zeroclaw_tool_call` host
//! function via Extism shared memory, and deserializes the JSON response.

use extism_pdk::*;
use serde::{Deserialize, Serialize};
use serde_json::Value;

// ---------------------------------------------------------------------------
// Request / response types (mirror the host-side structs)
// ---------------------------------------------------------------------------

#[derive(Debug, Serialize)]
struct ToolCallRequest {
    tool_name: String,
    arguments: Value,
}

#[derive(Debug, Deserialize)]
struct ToolCallResponse {
    success: bool,
    #[serde(default)]
    output: String,
    #[serde(default)]
    error: Option<String>,
}

// ---------------------------------------------------------------------------
// Host function imports
// ---------------------------------------------------------------------------

#[host_fn]
extern "ExtismHost" {
    fn zeroclaw_tool_call(input: Json<ToolCallRequest>) -> Json<ToolCallResponse>;
}

// ---------------------------------------------------------------------------
// Public wrapper API
// ---------------------------------------------------------------------------

/// Call a tool by name with the given JSON arguments.
pub fn tool_call(tool_name: &str, input: serde_json::Value) -> Result<String, Error> {
    let request = ToolCallRequest {
        tool_name: tool_name.to_string(),
        arguments: input,
    };
    let Json(response) = unsafe { zeroclaw_tool_call(Json(request))? };
    if let Some(err) = response.error {
        return Err(Error::msg(err));
    }
    if !response.success {
        return Err(Error::msg("tool call returned success=false"));
    }
    Ok(response.output)
}
